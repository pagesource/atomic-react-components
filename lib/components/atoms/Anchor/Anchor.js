// @flow
import React from 'react';
import type { Node } from 'react';
import styled from 'styled-components';

import { buildUrl } from '../../../core/utils/url';
import { getLabelFromChildren } from '../../../core/utils/elementLocator';
import styles from './Anchor.style';

type Props = {
  to: string | any,
  as?: string,
  children: Node,
  className?: string,
  elementLocator?: string,
  inheritedStyles?: string,
  handleLinkClick?: (SyntheticEvent<>) => void,
  shallow?: boolean,
  scroll?: boolean,
  target?: '' | '_self' | '_blank' | '_parent' | '_top',
  forceClick?: boolean,
  noLink: boolean,
};

const Anchor = ({
  children,
  to,
  as,
  className,
  elementLocator,
  inheritedStyles,
  handleLinkClick,
  shallow,
  scroll,
  forceClick,
  target,
  noLink,
  ...others
}: Props): Node => (
  <a
    href={buildUrl(to)}
    className={className}
    onClick={handleLinkClick}
    target={target}
    data-locator={elementLocator || `link-${getLabelFromChildren(children)}`}
    {...others}
  >
    {children}
  </a>
);

Anchor.defaultProps = {
  as: '',
  inheritedStyles: '',
  shallow: false,
  scroll: false,
  forceClick: false,
  className: '',
  elementLocator: '',
  target: '',
  handleLinkClick: () => {},
};

export default styled(Anchor)`
  ${styles};
`;

export { Anchor as AnchorVanilla };

export type { Props as AnchorProps };
