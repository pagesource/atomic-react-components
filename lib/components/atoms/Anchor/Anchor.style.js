import { css } from 'styled-components';

export default css`
  cursor: pointer;
  ${(props) => props.theme.behaviour.LINK_DECORATION};
  ${(props) => props.theme.colors.LINK_PRIMARY};
  ${(props) => props.theme.typography.FONT_LINK};
  &:focus {
    ${(props) => props.theme.border.CTA_OUTLINE};
    &:not(.focus-visible) {
      outline: none;
    }
  }
  ${(props) => (props.inheritedStyles ? props.inheritedStyles : '')};
`;
