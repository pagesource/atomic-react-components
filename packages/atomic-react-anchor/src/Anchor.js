import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { buildUrl, getLabelFromChildren } from '@xt-pagesource/atomic-react-utils';
import styles from './Anchor.style';

const Anchor = ({
  children,
  to,
  as,
  className,
  elementLocator,
  inheritedStyles,
  handleLinkClick,
  shallow,
  scroll,
  forceClick,
  target,
  ...others
}) => (
  <a
    {...others}
    href={buildUrl(to)}
    className={className}
    onClick={handleLinkClick}
    target={target}
    data-locator={elementLocator || `link-${getLabelFromChildren(children)}`}
  >
    {children}
  </a>
);

Anchor.propTypes = {
  to: PropTypes.string.isRequired,
  as: PropTypes.string,
  children: PropTypes.node.isRequired,
  className: PropTypes.string,
  elementLocator: PropTypes.string,
  inheritedStyles: PropTypes.string,
  handleLinkClick: PropTypes.func,
  shallow: PropTypes.bool,
  scroll: PropTypes.bool,
  target: PropTypes.oneOf(['', '_self', '_blank', '_parent', '_top']),
  forceClick: PropTypes.bool,
};

Anchor.defaultProps = {
  as: '',
  inheritedStyles: '',
  shallow: false,
  scroll: false,
  forceClick: false,
  className: '',
  elementLocator: '',
  target: '',
  handleLinkClick: () => {},
};

const StyledAnchor = styled(Anchor)`
  ${styles};
`;
export default StyledAnchor;

export { Anchor as AnchorVanilla };
