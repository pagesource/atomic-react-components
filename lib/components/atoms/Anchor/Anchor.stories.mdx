import { Meta, Story, Preview, Props  } from '@storybook/addon-docs/blocks';
import Anchor, { AnchorVanilla } from '.';
import { defaultAnchor, secondaryAnchor, customLocatorOnAnchor, openInNewTabAnchor } from './Anchor.mock';

<Meta title="Atoms/Links" component={Anchor} />


### Anchor Vanilla
<Preview>
  <Story name="Knobs">
   <AnchorVanilla {...defaultAnchor}>Click Me</AnchorVanilla>
  </Story>
</Preview>

### Secondary Anchor
<Preview>
    <Story name="Secondary Link">
  <Anchor {...secondaryAnchor}>Click Me</Anchor>
  </Story>
</Preview>


### Custom Locator On Anchor
<Preview>
    <Story name="Custom Locator On Link">
  <Anchor {...customLocatorOnAnchor}>Click Me</Anchor>
  </Story>
</Preview>

### Open In New Tab Anchor
<Preview>
    <Story name="Open In New Tab Link">
  <Anchor {...openInNewTabAnchor}>Click Me</Anchor>
  </Story>
</Preview>

## Props ðŸ”§

<Props of={AnchorVanilla} />

