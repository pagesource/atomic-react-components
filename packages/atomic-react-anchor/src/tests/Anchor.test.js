import React from 'react';
import { shallow, render } from 'enzyme';
import { AnchorVanilla } from '../Anchor';

describe('Anchor Component', () => {
  let anchor;

  test('should render correctly and match snapshot', () => {
    anchor = shallow(<AnchorVanilla to="/sample">Sample</AnchorVanilla>);
    expect(anchor).toMatchSnapshot();
  });

  test('Should handle on typeahead focuson Events', () => {
    const props = { noLink: true, to: '/sample', handleLinkClick: jest.fn() };
    const event = { preventDefault: jest.fn() };
    anchor = shallow(<AnchorVanilla {...props}>Sample</AnchorVanilla>);
    anchor.find('a').simulate('click', event);
    expect(props.handleLinkClick).toHaveBeenCalled();
  });

  test('should check the data attribute of anchor', () => {
    const sampleProps = { elementLocator: 'my-custom-anchor-locator' };
    const anchorVanilla = render(<AnchorVanilla {...sampleProps} />);

    expect(anchorVanilla.attr('data-locator')).toEqual(sampleProps.elementLocator);
  });
});
