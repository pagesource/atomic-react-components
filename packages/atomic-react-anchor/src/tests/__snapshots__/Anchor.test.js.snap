// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor Component should render correctly and match snapshot 1`] = `
<a
  className=""
  data-locator="link-sample"
  href="/sample"
  onClick={[Function]}
  target=""
>
  Sample
</a>
`;
